sgkit/stats/popgen.py

```diff
@@ -28,98 +31,168 @@ def diversity(
+        }
+    )
+    return conditional_merge_datasets(ds, new_ds, merge)
+
+
+# c = cohorts, k = alleles
+@guvectorize(  # type: ignore
+    ["void(int64[:, :], int64[:], float64[:,:])"],
+    "(c, k),(c)->(c,c)",
+    nopython=True,
+)
+def _divergence(ac: ArrayLike, an: ArrayLike, out: ArrayLike) -> None:
+    """Generalized U-function for computing divergence.
+
+    Parameters
+    ----------
+    ac
+        Allele counts of shape (cohorts, alleles) containing per-cohort allele counts.
+    an
+        Allele totals of shape (cohorts,) containing per-cohort allele totals.
+    out
+        Pairwise divergence stats with shape (cohorts, cohorts), where the entry at
+        (i, j) is the divergence between cohort i and cohort j.
+    """
+    out[:, :] = np.nan  # (cohorts, cohorts)
+    n_cohorts = ac.shape[0]
+    n_alleles = ac.shape[1]
+    # calculate the divergence for each cohort pair
+    for i in range(n_cohorts):
+        for j in range(i + 1, n_cohorts):
+            n_pairs = an[i] * an[j]
+            n_same = 0
+            for k in range(n_alleles):
+                n_same += ac[i, k] * ac[j, k]
+            n_diff = n_pairs - n_same
+            div = n_diff / n_pairs
+            out[i, j] = div
+            out[j, i] = div
 
 
 def divergence(
-    ds1: Dataset,
-    ds2: Dataset,
-    allele_counts: Hashable = "variant_allele_count",
-) -> DataArray:
-    """Compute divergence between two genotype call datasets.
+    ds: Dataset, allele_counts: Hashable = "cohort_allele_count", merge: bool = True
+) -> Dataset:
+    """Compute divergence between pairs of cohorts.
 
     Parameters
     ----------
-    ds1
-        Genotype call dataset.
-    ds2
+    ds
         Genotype call dataset.
     allele_counts
-        allele counts to use or calculate.
+        cohort allele counts to use or calculate.
 
     Returns
     -------
-    divergence value between the two datasets.
+    divergence value between pairs of cohorts.
     """
-    if allele_counts not in ds1:
-        ds1_new = count_variant_alleles(ds1)
-    else:
-        ds1_new = ds1
-    ac1 = ds1_new[allele_counts]
-    if allele_counts not in ds2:
-        ds2_new = count_variant_alleles(ds2)
-    else:
-        ds2_new = ds2
-    ac2 = ds2_new[allele_counts]
-    an1 = ds1_new[allele_counts].sum(axis=1)
-    an2 = ds2_new[allele_counts].sum(axis=1)
 
-    n_pairs = an1 * an2
-    n_same = (ac1 * ac2).sum(axis=1)
-    n_diff = n_pairs - n_same
-    div = n_diff / n_pairs
-    return div.sum()  # type: ignore[no-any-return]
+    if allele_counts not in ds:
+        ds = count_cohort_alleles(ds)
+    ac = ds[allele_counts]
+    an = ac.sum(axis=2)
+
+    n_variants = ds.dims["variants"]
+    n_cohorts = ds.dims["cohorts"]
+    ac = da.asarray(ac)
+    an = da.asarray(an)
+    shape = (ac.chunks[0], n_cohorts, n_cohorts)
+    d = da.map_blocks(_divergence, ac, an, chunks=shape, dtype=np.float64)
+    assert_array_shape(d, n_variants, n_cohorts, n_cohorts)
```

Comment:
👍  -- I'll try to remember to start doing that too after all `map_blocks` calls.